### ISSUE TITLE
Disable restoring unpinned dependencies in providers

### ISSUE THREAD

**Issue by danielrbradley** [pulumi/pulumi#11147]
Providers written for the `nodejs` runtime with a `package.json` will automatically have their dependencies restored by the engine running either `npm install` or `yarn install` as a post-install action. 

If no lock file is shipped in the provider package, then it's unknown which version of each dependency (including transient dependencies) will be installed. This could be exploited by a bad actor if they were able to take control of any of the unpinned sub-dependencies and publish a malicious version. This could be leveraged to allow arbitary code execution within the deployment environment.

Summary: We should only allow packages to be dynamically restored if there is a lock file present.

This can be implemented by using `yarn install --frozen-lockfile` or `npm ci` to restore packages.

**Warning**: this would break existing provider versions which don't include lock files.

**Comment by danielrbradley** [comment]
Implementation suggestion:

To avoid blocking users stuck on providers without lock files, we should only enable non-locked package restores by explicit opt-in.

Default behaviour:
- Use `npm ci` or `yarn install --frozen-lockfile` if there is a lock file present.
- Fail if there is no lock file present with an error with instructions on falling back to insecure behaviour (see below).

Provide a configuration option to explicitly enable non-deterministic package installations:
- Add configuration via environment variables (e.g. `PULUMI_ENABLE_INSECURE_PACKAGE_INSTALLATION`) or CLI flag (e.g. `--enable-insecure-package-installation`).

Additional considerations:
- Should we allow this flag to be set for only specific providers e.g. `PULUMI_ENABLE_INSECURE_PACKAGE_INSTALLATION=pulumi-eks`?
- Does the configuration name indicate clearly enough the risk and behaviour change?

**Comment by AaronFriel** [comment]
> Fail if there is no lock file present with an error with instructions on falling back to insecure behaviour (see below).

I think our breaking changes policy will prevent us from implementing a hard failure, but I definitely think we should warn when a lockfile is not present. The language hosts could detect this, and we could make it a hard error in a few releases.

**Comment by danielrbradley** [comment]
> I think our breaking changes policy will prevent us from implementing a hard failure

The breaking changes policy doesn't currently cover the issue where maintaining existing behaviour is a security vulnerability.

By the time a warning has been shown the issue could have already been exploited. We should however think about how we could mitigate user experience for existing package versions which we can't retrospectively update.

The user impact I'd expect the above fix to have is that all EKS users would be presented with an error on their next deployment which explains the risk and and how to set the required configuration to re-enable.

One way to make the change less disruptive would be to retrospecitvely ship corresponding lock files for popular EKS provider versions. I think this would be possible because we don't cache the checksums of provider binaries right now (also a possible issue to address in the future).

**Comment by Frassle** [comment]
> I think this would be possible because we don't cache the checksums of provider binaries right now (also a possible issue to address in the future).

Correct, I'm about half way done with implementing this but the checksums are set by the SDKs (not cached) so old versions would still be able to download updated provider plugins.

**Comment by AaronFriel** [comment]
> > I think our breaking changes policy will prevent us from implementing a hard failure
>
> The breaking changes policy doesn't currently cover the issue where maintaining existing behaviour is a security vulnerability.

I think this greatly underestimates the amount of customer harm done by forcing a principled perspective of security (which I agree with, mind you) on users who themselves have principled, even if we consider them wrong, oppositions to committing lockfiles.

**Comment by AaronFriel** [comment]
I don't think we can flag this as a security issue though. This is a supply chain security _enhancement_. Even if there were an active exploit here, the CVE would be with the dependency that is a victim of a supply chain attack, not Pulumi.

In the future, I expect US Executive Order 14017 and the subsequently drafted NIST publication SP 800-161 to consider issues like this a vulnerability. Like I said, I agree with the principled POV here. I just don't think we can ship a version of Pulumi that breaks customers suddenly.
